bokehjs/test/models/tools/edit/box_draw_tool.ts

```diff
@@ -62,50 +62,44 @@ describe("BoxDrawTool", () =>
       });
 
       this.hit_test_stub = sinon.stub(this.glyph_renderer_view.glyph, "hit_test");
-    });
 
-    it("should select rect on tap", function(): void {
-      this.hit_test_stub.returns(create_1d_hit_test_result([[1, 0]]));
-      const draw_tool = new BoxDrawTool({renderers: [this.glyph_renderer]});
+      const draw_tool = new BoxDrawTool({
+        active: true,
+        renderers: [this.glyph_renderer],
+        empty_value: "Test"
+      });
       this.plot.add_tools(draw_tool);
-      const draw_tool_view = this.plot_canvas_view.tool_views[draw_tool.id];
+      this.draw_tool_view = this.plot_canvas_view.tool_views[draw_tool.id];
```

Comment:
My preference is not to set on `this`, but to create a function that will spawn a new instance for every test. In fact, this code shouldn't compile, because implicit `this` for the enclosing function doesn't declare this attribute. Unless of course `mocha` is doing something nasty with types.

Reply:
Obviously, https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/mocha/index.d.ts#L139.

Reply:
I don't have polite words to describe this.

Reply:
>My preference is not to set on this, but to create a function that will spawn a new instance for every test.

Happy to do that, I'd have the function return an object containing a bunch of objects for easy access. The tools interacts with a lot of different models and views, which I need to access.

>I don't have polite words to describe this.

Nothing surprises me in the world of JS and its derivatives.

Reply:
> The tools interacts with a lot of different models and views, which I need to access.

You can define an interface for this purpose and return a literal object conforming to it. This is what you would do anyway to properly type the current approach, if `@types/mocha` didn't do its magic. You would have to have `function(this: mocha callback arg type & interface with your custom attributes)` everywhere. This is why I hate so much tricks that `@types/mocha` uses for users' convenience, because it tricked you into thinking that you're writing type safe code, whereas you were writing completely dynamically typed code. Note that `any` means anything can happen, and there is no type checking as soon as `any` appears in the code.

Reply:
Stylistic issues aside, not sharing state between tests is actually pretty important for correctness of tests, because you have to acknowledge that there are side effects of bokehjs' code you run. Unless you want to test those side effects, you should be always starting a test with a fresh set of objects. Of course, one can stub out certain classes of functionality to prohibit those side effects from happening, but stubbing code out isn't without its issues.

Reply:
>Stylistic issues aside, not sharing state between tests is actually pretty important for correctness of tests, because you have to acknowledge that there are side effects of bokehjs' code you run.

Isn't the point of the beforeEach block that it is rerun before each individual test which means none of the state is actually shared between tests?

Reply:
Still happy to split it out of course, if only to define proper types, just trying to clarify my understanding.